syntax = "proto3";

package search_ml;

option go_package = "./pb";

message Document {
    string id = 1;
    string title = 2;
    string content_snippet = 3;
}

message RankRequest {
    string query = 1;                       // search string
    repeated Document candidates = 2;       // n documents to reorder
}

message RankedDocument {
    string id = 1;
    float score = 2;                        // relevance score from 0 to 1
}

message RankResponse {
    repeated RankedDocument results = 1;    // n documents sorted by score.
}

message EmbeddingRequest {
    string text = 1;
}

message EmbeddingResponse {
    repeated float vector = 1;              // k dimensional embedding vector
}

service MLService {
    // Takes a query and documents, returns documents ranked by relevance.
    rpc ReRank (RankRequest) returns (RankResponse);
    // Convert query into embedding
    rpc GetEmbedding (EmbeddingRequest) returns (EmbeddingResponse);
}
